clevertap-core/src/main/java/com/clevertap/android/sdk/userEventLogs/UserEventLogDAOImpl.kt

```diff
@@ -0,0 +1,361 @@
+            emptyList()
+        }
+    }
+
+    // Create index on lastTs column if this method is frequently used
+    @WorkerThread
+    override fun allEvents(): List<UserEventLog> {
+        val tName = table.tableName
+        val eventList = mutableListOf<UserEventLog>()
+        val orderBy = "${Column.LAST_TS} ASC"
+
+        return try {
+            db.readableDatabase.query(
+                tName, null, null, null, null, null, orderBy
+            )?.use { cursor ->
+                while (cursor.moveToNext()) {
+                    val eventLog = UserEventLog(
+                        eventName = cursor.getString(cursor.getColumnIndexOrThrow(Column.EVENT_NAME)),
+                        firstTs = cursor.getLong(cursor.getColumnIndexOrThrow(Column.FIRST_TS)),
+                        lastTs = cursor.getLong(cursor.getColumnIndexOrThrow(Column.LAST_TS)),
+                        countOfEvents = cursor.getInt(cursor.getColumnIndexOrThrow(Column.COUNT)),
+                        deviceID = cursor.getString(cursor.getColumnIndexOrThrow(Column.DEVICE_ID))
+                    )
+                    eventList.add(eventLog)
+                }
+                eventList
+            } ?: emptyList()
+        } catch (e: Exception) {
+            logger.verbose("Could not fetch records out of database $tName.", e)
+            emptyList()
+        }
+    }
+
+    @WorkerThread
+    override fun cleanUpExtraEvents(threshold: Int, numberOfRowsToCleanup: Int): Boolean {
+        if (threshold <= 0) {
+            logger.verbose("Invalid threshold value: $threshold. Threshold should be greater than 0")
+            return false
+        }
+        if (numberOfRowsToCleanup < 0) {
+            logger.verbose("Invalid numberOfRowsToCleanup value: $numberOfRowsToCleanup. Should be greater than or equal to 0")
+            return false
+        }
+        if (numberOfRowsToCleanup >= threshold) {
+            logger.verbose("Invalid numberOfRowsToCleanup value: $numberOfRowsToCleanup. Should be less than threshold: $threshold")
+            return false
+        }
+
+        val tName = table.tableName
+        val numberOfRowsToKeep = threshold - numberOfRowsToCleanup
+
+        return try {
+            // SQL query to delete only the least recently used rows, using a subquery with LIMIT
+            // When above threshold is reached, delete in such a way that (threshold - numberOfRowsToCleanup) rows exists after cleanup
+            val query = """
+            DELETE FROM $tName
+            WHERE (${Column.EVENT_NAME}, ${Column.DEVICE_ID}) IN (
+                SELECT ${Column.EVENT_NAME}, ${Column.DEVICE_ID}
+                FROM $tName
+                ORDER BY ${Column.LAST_TS} ASC 
+                LIMIT (
+                SELECT CASE 
+                    WHEN COUNT(*) > ? THEN COUNT(*) - ?
+                    ELSE 0
+                END 
+                FROM $tName
+                )
+            );
+        """.trimIndent()
```

Comment:
extract sub-queries in different vals to make this readable?

Reply:
Let's keep as it is to cache compiled sql statement